core/src/main/java/cucumber/runtime/RuntimeGlue.java

```diff
@@ -60,19 +61,29 @@ public void addAfterHook(HookDefinition hookDefinition) {
 
     @Override
     public StepDefinitionMatch stepDefinitionMatch(String featurePath, PickleStep step) {
+        String stepText = step.getText();
+
+        CacheEntry cacheEntry = matchedStepDefinitionsCache.get(stepText);
+        if (cacheEntry != null) {
+            return new StepDefinitionMatch(cacheEntry.arguments, cacheEntry.stepDefinition, featurePath, step, localizedXStreams);
+        }
+
         List<StepDefinitionMatch> matches = stepDefinitionMatches(featurePath, step);
         if (matches.isEmpty()) {
             return null;
         }
         if (matches.size() == 1) {
-            return matches.get(0);
+            StepDefinitionMatch match = matches.get(0);
+            matchedStepDefinitionsCache.put(stepText, new CacheEntry(match.getStepDefinition(),match.getArguments()));
+            return match;
         } else {
             throw new AmbiguousStepDefinitionsException(step, matches);
```

Comment:
To stay consistent with the rest of this method it would be preferable to remove the else clause and simply throw the exception as the last statement in the method.